eth_account/hdaccount/mnemonic.py

```diff
@@ -0,0 +1,172 @@
+
+from eth_utils import (
+    ValidationError,
+    combomethod,
+)
+
+PBKDF2_ROUNDS = 2048
+VALID_SEED_SIZES = [16, 20, 24, 28, 32]
+VALID_WORD_LENGTHS = [12, 15, 18, 21, 24]
+WORDLIST_DIR = Path(__file__).parent / "wordlist"
+
+
+class ConfigurationError(Exception):
+    pass
+
+
+def normalize_string(txt):
+    if isinstance(txt, bytes):
+        utxt = txt.decode("utf8")
+    elif isinstance(txt, str):
+        utxt = txt
+    else:
+        raise TypeError("String value expected")
+
+    return unicodedata.normalize("NFKD", utxt)
+
+
+class Mnemonic(object):
+    def __init__(self, language):
+        if language not in self.list_languages():
+            raise ValidationError(
+                f'Invalid language choice "{language}", must be one of {self.list_langauges()}'
+            )
+        self.language = language
+        self.radix = 2048
+        with open(WORDLIST_DIR / f"{language}.txt", "r", encoding="utf-8") as f:
+            self.wordlist = [w.strip() for w in f.readlines()]
+        if len(self.wordlist) != self.radix:
+            raise ConfigurationError(
+                f"Wordlist should contain {self.radix} words, "
+                f"but it contains {len(self.wordlist)} words."
+            )
+
+    @combomethod
+    def list_languages(_Mnemonic):
+        return sorted(Path(f).stem for f in WORDLIST_DIR.rglob("*.txt"))
+
+    @classmethod
+    def detect_language(cls, raw_mnemonic):
+        mnemonic = normalize_string(raw_mnemonic)
+        first = mnemonic.split(" ")[0]
+        languages = cls.list_languages()
+
+        for lang in languages:
+            if first in cls(lang).wordlist:
+                return lang
+
+        raise ConfigurationError("Language not detected")
+
+    def generate(self, num_words=12):
+        if num_words not in VALID_WORD_LENGTHS:
+            raise ValueError(
+                f"Invalid choice for number of words: {num_words}, should be one of "
+                f"{VALID_WORD_LENGTHS}"
+            )
+        return self.to_mnemonic(os.urandom(4 * num_words // 3))  # 4/3 bytes per word
+
+    def to_mnemonic(self, seed):
+        if len(seed) not in VALID_SEED_SIZES:
+            raise ValueError(
+                f"Invalid data length {len(seed)}, should be one of "
+                f"{VALID_WORD_LENGTHS}"
+            )
+        checksum = hashlib.sha256(seed).hexdigest()
+        bits = (
+            bin(int(binascii.hexlify(seed), 16))[2:].zfill(len(seed) * 8) +
+            bin(int(checksum, 16))[2:].zfill(256)[: len(seed) * 8 // 32]
+        )
+        result = []
+        for i in range(len(bits) // 11):
+            idx = int(bits[i * 11: (i + 1) * 11], 2)
+            result.append(self.wordlist[idx])
+        if self.language == "japanese":  # Japanese must be joined by ideographic space.
+            result_phrase = u"\u3000".join(result)
+        else:
+            result_phrase = " ".join(result)
+        return result_phrase
+
+    def check(self, mnemonic):
```

Comment:
Is this one that you're going to use later? We should maybe add some testing if so. If not, I'd vote to remove, but could be convinced otherwise.

Reply:
Yes, so check is used with #87 to ensure the checksum validates before using the seed phrase (the checksum is not necessary for seed generation, per BIP39)